pyop2/fusion.py

```diff
@@ -0,0 +1,1264 @@
+#       permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTERS
+# ''AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+# OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""OP2 backend for fusion and tiling of ``ParLoops``."""
+
+from contextlib import contextmanager
+from collections import OrderedDict
+from copy import deepcopy as dcopy, copy as scopy
+import os
+
+from base import *
+import base
+import compilation
+import host
+from backends import _make_object
+from caching import Cached
+from profiling import lineprof, timed_region, profile
+from logger import warning, info as log_info
+from mpi import collective
+from configuration import configuration
+from utils import flatten, strip, as_tuple
+
+import coffee
+from coffee import base as ast
+from coffee.plan import ASTKernel
+from coffee.utils import visit as ast_visit, ast_c_make_alias as ast_make_alias
+
+import slope_python as slope
+
+
+class Arg(host.Arg):
+
+    @staticmethod
+    def specialize(args, gtl_map, loop_id):
+        """Given ``args``, instances of some :class:`fusion.Arg` superclass,
+        create and return specialized :class:`fusion.Arg` objects.
+
+        :param args: either a single :class:`host.Arg` object or an iterator
+                     (accepted: list, tuple) of :class:`host.Arg` objects.
+        :gtl_map: a dict associating global maps' names to local maps' c_names.
+        :param loop_id: indicates the position of the args` loop in the loop
+                        chain
+        """
+
+        def convert(arg, gtl_map, loop_id):
+            # Retrive local maps
+            maps = as_tuple(arg.map, Map)
+            c_local_maps = [None]*len(maps)
+            for i, map in enumerate(maps):
+                c_local_maps[i] = [None]*len(map)
+                for j, m in enumerate(map):
+                    c_local_maps[i][j] = gtl_map["%s%d_%d" % (m.name, i, j)]
+            # Instantiate and initialize new, specialized Arg
+            _arg = Arg(arg.data, arg.map, arg.idx, arg.access, arg._flatten)
+            _arg._loop_position = loop_id
+            _arg._position = arg._position
+            _arg._indirect_position = arg._indirect_position
+            _arg._c_local_maps = c_local_maps
+            return _arg
+
+        if isinstance(args, (list, tuple)):
```

Comment:
Bonus question: is this code ever called with a list or a tuple? The only use I could find already loops over a set of elements outside the call.